libcaf_io/src/default_multiplexer.cpp

```diff
@@ -279,19 +280,159 @@ namespace network {
+#if defined(CAF_KQUEUE_MULTIPLEXER)
+
+  default_multiplexer::default_multiplexer(actor_system* sys)
+      : multiplexer(sys),
+        loopfd_(invalid_native_socket),
+        pipe_reader_(*this) {
+    init();
+    // initialize kqueue
+    loopfd_ = kqueue();
+    if (loopfd_ == -1) {
+      CAF_LOG_ERROR("kqueue: " << strerror(errno));
+      exit(errno);
+    }
+    // add pipe to poll set
+    pipe_ = create_pipe();
+    pipe_reader_.init(pipe_.first);
+    pollset_.emplace_back();
+    auto ke = &pollset_.back();
+    EV_SET(ke, pipe_reader_.fd(), EVFILT_READ, EV_ADD, 0, 0, &pipe_reader_);
+    if (kevent(loopfd_, ke, 1, nullptr, 0, nullptr) < 0) {
+      CAF_LOG_ERROR("kqueue: " << strerror(errno));
+      CAF_CRITICAL("kqueue() failed");
+    }
+    shadow_.fds.push_back(pipe_reader_.fd());
+  }
+
+  void default_multiplexer::run() {
+    CAF_LOG_TRACE("kqueue()-based multiplexer");
+    pollset_.resize(20);
+    while (!shadow_.fds.empty()) {
+      int nev = kevent(loopfd_, shadow_.changes.data(),
+                       static_cast<int>(shadow_.changes.size()),
+                       pollset_.data(), static_cast<int>(pollset_.size()),
+                       nullptr);
+      CAF_LOG_DEBUG("kevent() on " << CAF_ARG(pollset_.size())
+                    << " sockets reported " << CAF_ARG(nev)
+                    << " event(s)");
+      if (nev < 0)
+        switch (errno) {
+          case EINTR:
+            shadow_.changes.clear(); // all changes guaranteed to be applied
+            continue;
+          default:
+            CAF_LOG_ERROR("kevent: " << strerror(errno));
+        }
+      shadow_.changes.clear();
+      auto iter = pollset_.begin();
+      auto last = iter + nev;
+      for (; iter != last; ++iter) {
+        auto fd = static_cast<native_socket>(iter->ident);
+        auto ptr = reinterpret_cast<event_handler*>(iter->udata);
+        CAF_ASSERT(ptr != nullptr);
+        if (iter->flags & EV_ERROR || iter->flags & EV_EOF) {
+          CAF_LOG_DEBUG("error occured on socket:"
+                        << CAF_ARG(fd) << CAF_ARG(iter->data)
+                        << CAF_ARG(strerror(iter->data)));
+          auto i = std::lower_bound(shadow_.fds.begin(), shadow_.fds.end(), fd);
+          if (i != shadow_.fds.end() && *i == fd) {
+            ptr->handle_event(operation::propagate_error);
+            shadow_.fds.erase(i);
+          }
+          // Pending events for the failed handler are no longer relevant.
+          auto e = std::lower_bound(events_.begin(), events_.end(), fd,
+                                    event_less{});
+          while (e != events_.end() && e->fd == fd)
+            e = events_.erase(e);
+        } else if (iter->filter == EVFILT_READ && !ptr->read_channel_closed()) {
+          ptr->handle_event(operation::read);
+        } else if (iter->filter == EVFILT_WRITE) {
+          ptr->handle_event(operation::write);
+        } else {
+          CAF_ASSERT(!"unexpected kevent filter");
```

Comment:
I don't like abusing `CAF_ASSERT` for reporting hard errors. Please use `CAF_CRITICAL` instead.